File: examples/go/process-manager-fulfillment/logic/pm_logic.go
// Package logic implements the order-fulfillment process manager: a fan-in
// coordinator that tracks three prerequisites (payment, inventory, packing)
// and issues a Ship command to the fulfillment domain when all are complete.
//
// Internal events (PrerequisiteCompleted, DispatchIssued) are stored as
// JSON-encoded anypb.Any values in process manager state for replay.
package logic

import (
	"encoding/json"
	"fmt"
	"strings"

	angzarrpb "angzarr/proto/angzarr"
	"angzarr/proto/examples"

	"github.com/google/uuid"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

const (
	PMName   = "order-fulfillment"
	PMDomain = "order-fulfillment"

	fulfillmentDomain = "fulfillment"

	prereqPayment     = "payment"
	prereqInventory   = "inventory"
	prereqFulfillment = "fulfillment"

	typeURLPrerequisiteCompleted = "type.examples/examples.PrerequisiteCompleted"
	typeURLDispatchIssued        = "type.examples/examples.DispatchIssued"

	dispatchedMarker = "__dispatched__"
)

var allPrerequisites = []string{prereqPayment, prereqInventory, prereqFulfillment}

// prerequisiteCompleted is stored in PM state when a prerequisite is satisfied.
type prerequisiteCompleted struct {
	Prerequisite string   `json:"prerequisite"`
	Completed    []string `json:"completed"`
	Remaining    []string `json:"remaining"`
}

// dispatchIssued is stored in PM state when all prerequisites are met.
type dispatchIssued struct {
	Completed []string `json:"completed"`
}

// Handle is the PMHandleFunc for the order-fulfillment process manager.
//
// It classifies the trigger event, replays process state to determine which
// prerequisites are already satisfied, and emits a Ship command when all three
// prerequisites are complete. DispatchIssued in state prevents duplicate dispatch.
func Handle(trigger *angzarrpb.EventBook, processState *angzarrpb.EventBook, _ []*angzarrpb.EventBook) ([]*angzarrpb.CommandBook, *angzarrpb.EventBook) {
	correlationID := ""
	if c := trigger.GetCover(); c != nil {
		correlationID = c.CorrelationId
	}
	if correlationID == "" {
		return nil, nil
	}

	completed := extractCompleted(processState)

	if alreadyDispatched(completed) {
		return nil, nil
	}

	var newPrereq string
	for _, page := range trigger.GetPages() {
		if page.GetEvent() == nil {
			continue
		}
		prereq := classifyEvent(page.GetEvent())
		if prereq == "" {
			continue
		}
		if !contains(completed, prereq) {
			completed = append(completed, prereq)
			newPrereq = prereq
		}
	}

	if newPrereq == "" {
		return nil, nil
	}

	pmRootUUID := uuid.NewSHA1(uuid.NameSpaceOID, []byte(correlationID))
	pmRoot := &angzarrpb.UUID{
		Value: pmRootUUID[:],
	}

	nextSeq := nextSequence(processState)

	var pmPages []*angzarrpb.EventPage
	var commands []*angzarrpb.CommandBook

	remaining := difference(allPrerequisites, completed)
	prereqEvt := prerequisiteCompleted{
		Prerequisite: newPrereq,
		Completed:    copyStrings(completed),
		Remaining:    remaining,
	}
	prereqBytes, _ := json.Marshal(prereqEvt)

	pmPages = append(pmPages, &angzarrpb.EventPage{
		Sequence:  &angzarrpb.EventPage_Num{Num: nextSeq},
		CreatedAt: timestamppb.Now(),
		Event: &anypb.Any{
			TypeUrl: typeURLPrerequisiteCompleted,
			Value:   prereqBytes,
		},
	})

	if allComplete(completed) {
		dispatchEvt := dispatchIssued{
			Completed: copyStrings(completed),
		}
		dispatchBytes, _ := json.Marshal(dispatchEvt)

		pmPages = append(pmPages, &angzarrpb.EventPage{
			Sequence:  &angzarrpb.EventPage_Num{Num: nextSeq + 1},
			CreatedAt: timestamppb.Now(),
			Event: &anypb.Any{
				TypeUrl: typeURLDispatchIssued,
				Value:   dispatchBytes,
			},
		})

		orderID := rootIDAsString(trigger.GetCover().GetRoot())

		shipCmd := &examples.Ship{
			Carrier:        fmt.Sprintf("auto-%s", orderID),
			TrackingNumber: "",
		}
		shipAny, err := anypb.New(shipCmd)
		if err != nil {
			return nil, nil
		}

		commands = append(commands, &angzarrpb.CommandBook{
			Cover: &angzarrpb.Cover{
				Domain:        fulfillmentDomain,
				Root:          trigger.GetCover().GetRoot(),
				CorrelationId: correlationID,
			},
			Pages: []*angzarrpb.CommandPage{
				{Command: shipAny},
			},
		})
	}

	pmEventBook := &angzarrpb.EventBook{
		Cover: &angzarrpb.Cover{
			Domain:        PMDomain,
			Root:          pmRoot,
			CorrelationId: correlationID,
		},
		Pages: pmPages,
	}

	return commands, pmEventBook
}

// classifyEvent maps a trigger event type_url to its prerequisite name.
func classifyEvent(event *anypb.Any) string {
	url := event.GetTypeUrl()
	switch {
	case strings.HasSuffix(url, "PaymentSubmitted"):
		return prereqPayment
	case strings.HasSuffix(url, "StockReserved"):
		return prereqInventory
	case strings.HasSuffix(url, "ItemsPacked"):
		return prereqFulfillment
	default:
		return ""
	}
}

// extractCompleted replays process state pages and returns the set of completed
// prerequisite names. If DispatchIssued is found, the dispatchedMarker is included.
func extractCompleted(processState *angzarrpb.EventBook) []string {
	if processState == nil {
		return nil
	}

	var completed []string
	for _, page := range processState.GetPages() {
		event := page.GetEvent()
		if event == nil {
			continue
		}
		url := event.GetTypeUrl()
		if strings.HasSuffix(url, "PrerequisiteCompleted") {
			var evt prerequisiteCompleted
			if err := json.Unmarshal(event.Value, &evt); err != nil {
				continue
			}
			if !contains(completed, evt.Prerequisite) {
				completed = append(completed, evt.Prerequisite)
			}
		} else if strings.HasSuffix(url, "DispatchIssued") {
			if !contains(completed, dispatchedMarker) {
				completed = append(completed, dispatchedMarker)
			}
		}
	}
	return completed
}

// allComplete returns true when all three prerequisites are present.
func allComplete(completed []string) bool {
	for _, p := range allPrerequisites {
		if !contains(completed, p) {
			return false
		}
	}
	return true
}

// alreadyDispatched returns true when the dispatchedMarker is present.
func alreadyDispatched(completed []string) bool {
	return contains(completed, dispatchedMarker)
}

// nextSequence returns the next sequence number for a new PM event page.
func nextSequence(processState *angzarrpb.EventBook) uint32 {
	if processState == nil {
		return 0
	}
	pages := processState.GetPages()
	if len(pages) == 0 {
		return 0
	}
	last := pages[len(pages)-1]
	if num, ok := last.GetSequence().(*angzarrpb.EventPage_Num); ok {
		return num.Num + 1
	}
	return 0
}

// rootIDAsString converts a proto UUID to a human-readable string (or "unknown").
func rootIDAsString(root *angzarrpb.UUID) string {
	if root == nil || len(root.Value) != 16 {
		return "unknown"
	}
	u, err := uuid.FromBytes(root.Value)
	if err != nil {
		return "unknown"
	}
	return u.String()
}

// contains checks if a string slice includes a value.
func contains(slice []string, val string) bool {
	for _, s := range slice {
		if s == val {
			return true
		}
	}
	return false
}

// difference returns elements in all that are not in completed.
func difference(all, completed []string) []string {
	var result []string
	for _, a := range all {
		if !contains(completed, a) {
			result = append(result, a)
		}
	}
	return result
}

// copyStrings returns a shallow copy of a string slice.
func copyStrings(s []string) []string {
	out := make([]string, len(s))
	copy(out, s)
	return out
}

File: examples/python/pmg-fulfillment/pm_logic.py
"""Order Fulfillment Process Manager - fan-in across order, inventory, and fulfillment domains.

Demonstrates the fan-in pattern: action triggers only when ALL THREE domains
have completed their part. A saga cannot handle this because each saga instance
only sees one domain's event, and race conditions prevent reliable "all complete"
detection.

The Process Manager solves this by maintaining event-sourced state tracking
completed prerequisites, serializing concurrent updates via aggregate sequence,
and using a dispatch_issued flag for exactly-once command dispatch.

Subscribed Domains:
  - order: Listens for PaymentSubmitted
  - inventory: Listens for StockReserved
  - fulfillment: Listens for ItemsPacked

When all three events arrive (any order), emits a Ship command to fulfillment.
"""

from __future__ import annotations

import json
import uuid

from google.protobuf.any_pb2 import Any

from angzarr import types_pb2 as types
from proto import fulfillment_pb2 as fulfillment

PM_NAME = "pmg-fulfillment"
PM_DOMAIN = "pmg-fulfillment"

FULFILLMENT_DOMAIN = "fulfillment"

PREREQ_PAYMENT = "payment"
PREREQ_INVENTORY = "inventory"
PREREQ_FULFILLMENT = "fulfillment"

ALL_PREREQUISITES = [PREREQ_PAYMENT, PREREQ_INVENTORY, PREREQ_FULFILLMENT]

_DISPATCHED_MARKER = "__dispatched__"


def _extract_completed(process_state: types.EventBook) -> list[str]:
    """Extract completed prerequisites from process manager state events."""
    completed: list[str] = []
    for page in process_state.pages:
        event = page.event
        if not event.type_url:
            continue
        if event.type_url.endswith("PrerequisiteCompleted"):
            data = json.loads(event.value)
            prereq = data["prerequisite"]
            if prereq not in completed:
                completed.append(prereq)
        elif event.type_url.endswith("DispatchIssued"):
            completed.append(_DISPATCHED_MARKER)
    return completed


def _classify_event(event: Any) -> str | None:
    """Classify a trigger event into a prerequisite name."""
    if event.type_url.endswith("PaymentSubmitted"):
        return PREREQ_PAYMENT
    if event.type_url.endswith("StockReserved"):
        return PREREQ_INVENTORY
    if event.type_url.endswith("ItemsPacked"):
        return PREREQ_FULFILLMENT
    return None


def _all_complete(completed: list[str]) -> bool:
    """Check if all prerequisites are met."""
    return all(p in completed for p in ALL_PREREQUISITES)


def _already_dispatched(completed: list[str]) -> bool:
    """Check if dispatch was already issued (idempotency)."""
    return _DISPATCHED_MARKER in completed


def _root_id_as_string(root: types.UUID | None) -> str:
    """Convert a proto UUID to a hex string for display."""
    if root is None or not root.value:
        return ""
    return uuid.UUID(bytes=bytes(root.value)).hex


def handle(
    trigger: types.EventBook,
    process_state: types.EventBook,
    destinations: list[types.EventBook],
) -> tuple[list[types.CommandBook], types.EventBook | None]:
    """Handle a trigger event, returning commands and optional PM events.

    Tracks three prerequisites across domains. When all are met, issues
    a Ship command to the fulfillment domain.
    """
    correlation_id = trigger.cover.correlation_id if trigger.cover else ""
    if not correlation_id:
        return ([], None)

    # Get current completed prerequisites from PM state
    completed = _extract_completed(process_state)

    # Already dispatched - idempotent no-op
    if _already_dispatched(completed):
        return ([], None)

    # Classify the trigger event
    new_prerequisite = None
    for page in trigger.pages:
        event = page.event
        if not event.type_url:
            continue
        prereq = _classify_event(event)
        if prereq is not None and prereq not in completed:
            completed.append(prereq)
            new_prerequisite = prereq

    # No new prerequisite from this event
    if new_prerequisite is None:
        return ([], None)

    # Derive deterministic PM root from correlation_id
    pm_root = types.UUID(
        value=uuid.uuid5(uuid.NAMESPACE_OID, correlation_id).bytes,
    )

    # Compute next sequence from existing state
    next_seq = 0
    if process_state.pages:
        last_page = process_state.pages[-1]
        if last_page.HasField("sequence") and last_page.WhichOneof("sequence") == "num":
            next_seq = last_page.num + 1

    pm_pages: list[types.EventPage] = []
    commands: list[types.CommandBook] = []

    # Record prerequisite completion
    remaining = [p for p in ALL_PREREQUISITES if p not in completed]
    prereq_data = json.dumps({
        "prerequisite": new_prerequisite,
        "completed": list(completed),
        "remaining": remaining,
    }).encode()

    pm_pages.append(types.EventPage(
        num=next_seq,
        event=Any(
            type_url="type.examples/examples.PrerequisiteCompleted",
            value=prereq_data,
        ),
    ))

    # Check if all prerequisites met
    if _all_complete(completed):
        dispatch_data = json.dumps({
            "completed": list(completed),
        }).encode()

        pm_pages.append(types.EventPage(
            num=next_seq + 1,
            event=Any(
                type_url="type.examples/examples.DispatchIssued",
                value=dispatch_data,
            ),
        ))

        # Emit Ship command to fulfillment domain
        order_id = _root_id_as_string(
            trigger.cover.root if trigger.cover else None,
        )
        ship_cmd = fulfillment.Ship(
            carrier=f"auto-{order_id}",
            tracking_number="",
        )
        cmd_any = Any()
        cmd_any.Pack(ship_cmd, type_url_prefix="type.examples/")

        commands.append(types.CommandBook(
            cover=types.Cover(
                domain=FULFILLMENT_DOMAIN,
                root=trigger.cover.root if trigger.cover else None,
                correlation_id=correlation_id,
            ),
            pages=[types.CommandPage(command=cmd_any)],
        ))

    pm_event_book = types.EventBook(
        cover=types.Cover(
            domain=PM_DOMAIN,
            root=pm_root,
            correlation_id=correlation_id,
        ),
        pages=pm_pages,
    )

    return (commands, pm_event_book)
